src/compile/data/data.ts

```diff
@@ -79,61 +83,151 @@ export interface SummaryComponent {
+  }
+
+  return dataComponent;
 }
 
+function mergeChildren(model: Model, dataComponent: DataComponent, children: UnitModel[]): DataComponent {
+  if (children.length === 0) {
+    return dataComponent;
+  }
+
+  const childDataComponents = children.map((child: UnitModel) => child.component.data);
+
+  // set the parent source as child source
+  childDataComponents.forEach((data) => {
+    if (data.source === undefined) {
+      // child does not define its own source
+      data.source = {
+        source: model.dataName(SOURCE)
+      };
+    } else {
+      data.source.source = model.dataName(SOURCE);
+    }
+    delete data.source.format;
+    delete data.source.url;
+  });
+
+  formatParse.merge(dataComponent, childDataComponents);
+
+  const mergedAllFormulas = calculate.merge(dataComponent, childDataComponents);
+
+  if (!mergedAllFormulas) {
+    // we could not merge all formula up so we have to stop merging because other
+    // transforms depend on this
+    return dataComponent;
+  }
+
+  timeUnit.merge(dataComponent, childDataComponents);
+
+  const sameFilter = filter.mergeIfEqual(dataComponent, childDataComponents);
+  const compatibleNullFilter = nullFilter.mergeIfCompatible(dataComponent, childDataComponents);
+
+  if (!sameFilter || !compatibleNullFilter) {
+    return dataComponent;
+  }
+
+  // merge up bin if the children have no nullfilter or filter defined
+  // and the binning is the same
+  bin.merge(dataComponent, childDataComponents);
+
+  // TODO: selectionfilter
+
+  aggregate.merge(dataComponent, childDataComponents);
+  colorRank.merge(dataComponent, childDataComponents);
+
+  nonPositiveFilter.mergeIfCompatible(dataComponent, childDataComponents);
+
+  return dataComponent;
+}
+
+/**
+ * Merges data from children up if possible.
+ */
 export function parseLayerData(model: LayerModel): DataComponent {
-  return {
-    // filter and formatParse could cause us to not be able to merge into parent
-    // so let's parse them first
-    filter: filter.parseLayer(model),
-    formatParse: formatParse.parseLayer(model),
-    nullFilter: nullFilter.parseLayer(model),
-    nonPositiveFilter: nonPositiveFilter.parseLayer(model),
-
-    // everything after here does not affect whether we can merge child data into parent or not
-    source: source.parseLayer(model),
-    bin: bin.parseLayer(model),
-    calculate: formula.parseLayer(model),
-    timeUnit: timeUnit.parseLayer(model),
-    timeUnitDomain: timeUnitDomain.parseLayer(model),
-    summary: summary.parseLayer(model),
-    stackScale: stackScale.parseLayer(model),
-    colorRank: colorRank.parseLayer(model)
-  };
+  const dataComponent = parseData(model);
+
+  // function to compare the data sources
+  function dataHash(component: DataComponent) {
+    const source = component.source;
+    if (source.values) {
+      return hash(source.values);
+    }
+    return hash(source.url + source.format);
+  }
+
+  // move time unit domain by moving it up
```

Comment:
Why is time unit domain moved up in `parseLayerData`, but not in `parseFacetData`.